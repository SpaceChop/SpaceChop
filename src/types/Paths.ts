import * as t from 'runtypes';

// Test to make sure :preset exists in path.
const pattern = /^\/.*:preset(\([^)]+\))?([^\w)]*?$|\/.*)/i;

const Paths = t.Array(t.String
  .withConstraint((n) => !!n || 'Cannot be empty')
  .withConstraint((n) => /^\//.test(n) || 'Must start with /')
  .withConstraint((n) => pattern.test(n) || 'Requires :preset in path'),
).withConstraint((n) => n && n.length > 0 || 'Requires at least one path');

export type Paths = t.Static<typeof Paths>;
export default Paths;
