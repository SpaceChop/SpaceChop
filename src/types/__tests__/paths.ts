import Paths from '../Paths';

describe('validate Paths', () => {
  describe('invalid', () => {
    it('should require paths', () => {
      const paths = null;
      expect(() => Paths.check(paths)).toThrowErrorMatchingSnapshot();
    });

    it('should throw on invalid paths array', () => {
      let paths = {};
      expect(() => Paths.check(paths)).toThrowErrorMatchingSnapshot();
      paths = [];
      expect(() => Paths.check(paths)).toThrowErrorMatchingSnapshot();
      paths = [''];
      expect(() => Paths.check(paths)).toThrowErrorMatchingSnapshot();
      paths = ['/'];
      expect(() => Paths.check(paths)).toThrowErrorMatchingSnapshot();
      paths = ['/:image'];
      expect(() => Paths.check(paths)).toThrowErrorMatchingSnapshot();
      paths = [':preset'];
      expect(() => Paths.check(paths)).toThrowErrorMatchingSnapshot();
    });

    it('should not throw on non capture groups in path', () => {
      const paths = ['/:preset/:image([^.]+).(.*)'];
      expect(() => Paths.check(paths)).not.toThrow();
    });
  });
});
