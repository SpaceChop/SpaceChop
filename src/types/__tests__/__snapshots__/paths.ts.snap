// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate paths should require paths 1`] = `"paths: Expected array, but was undefined"`;

exports[`validate paths should throw on invalid paths array 1`] = `"paths: Expected array, but was object"`;

exports[`validate paths should throw on invalid paths array 2`] = `"paths: Requires at least one path"`;

exports[`validate paths should throw on invalid paths array 3`] = `"paths.[0]: Cannot be empty"`;

exports[`validate paths should throw on invalid paths array 4`] = `"paths.[0]: Requires :preset in path"`;

exports[`validate paths should throw on invalid paths array 5`] = `"paths.[0]: Requires :preset in path"`;

exports[`validate paths should throw on invalid paths array 6`] = `"paths.[0]: Must start with /"`;

exports[`validate paths should throw on missing path params in config 1`] = `"params.width: Could not find any usage in sources, storage or paths"`;

exports[`validate paths should throw on missing path params in config 2`] = `"params.width: Could not find any usage in sources, storage or paths"`;

exports[`validate paths should throw on missing path params in config 3`] = `"params.width: Could not find any usage in sources, storage or paths"`;

exports[`validate paths should throw on missing path params in config 4`] = `"params.image: Could not find any usage in sources, storage or paths"`;

exports[`validate paths should throw on missing path params in config 5`] = `"params.height: Could not find any usage in sources, storage or paths"`;

exports[`validate paths should throw on missing path params in config 6`] = `"presets.t_200.steps[0].width: Missing param (width) in paths for preset (t_200)"`;

exports[`validate paths should throw on missing path params in config 7`] = `"source[0].volume.root: Missing param \\"image\\" in paths"`;
